solr/core/src/java/org/apache/solr/util/SafeXMLParsing.java

```diff
@@ -51,36 +51,51 @@ public final class SafeXMLParsing {
   private SafeXMLParsing() {}
 
   /**
-   * Parses a config file from ResourceLoader. Xinclude and external entities are enabled, but
-   * cannot escape the resource loader.
+   * Parses a config file from a Solr config based on ResourceLoader. Xinclude and external entities
+   * are enabled, but cannot escape the resource loader.
    */
   public static Document parseConfigXML(Logger log, ResourceLoader loader, String file)
       throws SAXException, IOException {
     try (InputStream in = loader.openResource(file)) {
+      DocumentBuilder db = configDocumentBuilder(loader, log);
+      return db.parse(in, SystemIdResolver.createSystemIdFromResourceName(file));
+    }
+  }
+
+  /**
+   * Parses a config file from a Solr config based on InputSource. Xinclude and external entities
+   * are enabled, but cannot escape the resource loader.
+   */
+  public static Document parseConfigXML(Logger log, InputSource is, ResourceLoader loader)
+      throws SAXException, IOException {
+    return configDocumentBuilder(loader, log).parse(is);
+  }
+
+  private static DocumentBuilder configDocumentBuilder(ResourceLoader loader, Logger log) {
+    try {
+      if (null == loader) throw new NullPointerException("loader");
       final DocumentBuilderFactory dbf = DocumentBuilderFactory.newInstance();
-      dbf.setValidating(false);
       dbf.setNamespaceAware(true);
       trySetDOMFeature(dbf, XMLConstants.FEATURE_SECURE_PROCESSING, true);
-      try {
-        dbf.setXIncludeAware(true);
-      } catch (UnsupportedOperationException e) {
-        throw new SolrException(
-            SolrException.ErrorCode.BAD_REQUEST, "XML parser doesn't support XInclude option", e);
-      }
-
+      // only enable xinclude, if systemId is available. this assumes it is the case as loader
+      // provides one.
+      dbf.setXIncludeAware(true);
```

Comment:
I like how before it was clear that setXIncludeAware might throw an UnsupportedOperationException.  Now it isn't.  You could ameliorate that simply by adding `// may throw UnsupportedOperationException` at the end.

I suppose it's extremely unlikely such an exception would be thrown by other stuff in the broader try-finally.  I've never seen this any way so it's a minor matter.